crates/verkle/db/src/verkle_db.rs

```diff
@@ -0,0 +1,320 @@
+    pub cache: MemoryDb,
+}
+
+impl<'a, TX> VerkleDb<'a, TX>{
+/// Create new [`VerkleDb`].
+pub fn new(tx: &'a TX) -> Self {
+    Self { tx, batch: MemoryDb::new(), cache: MemoryDb::new() }
+}
+}
+
+impl<'a, TX: DbTxMut + DbTx + Copy> Flush for VerkleDb<'a, TX> 
+{
+    fn flush(&mut self) {
+        let now = std::time::Instant::now();
+        let mut trie_updates: HashMap<Vec<u8>, Vec<u8>> = HashMap::new();
+
+        // Process leaf table
+        for (key, value) in self.batch.leaf_table.iter() {
+            let mut labelled_key = Vec::with_capacity(key.len() + 1);
+            labelled_key.push(LEAF_TABLE_MARKER);
+            labelled_key.extend_from_slice(key);
+            trie_updates.insert(labelled_key, value.to_vec());
+        }
+
+        // Process stem table
+        for (key, meta) in self.batch.stem_table.iter() {
+            let mut labelled_key = Vec::with_capacity(key.len() + 1);
+            labelled_key.push(STEM_TABLE_MARKER);
+            labelled_key.extend_from_slice(key);
+            trie_updates.insert(labelled_key, meta.to_bytes().unwrap());
+        }
+
+        // Process branch table
+        for (branch_id, b_child) in self.batch.branch_table.iter() {
+            let mut labelled_key = Vec::with_capacity(branch_id.len() + 1);
+            labelled_key.push(BRANCH_TABLE_MARKER);
+            labelled_key.extend_from_slice(branch_id);
+
+            match b_child {
+                BranchChild::Stem(stem_id) => {
+                    trie_updates.insert(labelled_key, stem_id.to_vec());
+                }
+                BranchChild::Branch(b_meta) => {
+                    trie_updates.insert(labelled_key, b_meta.to_bytes().unwrap());
+                }
+            }
+        }
+
+        for (key, value) in trie_updates {
+            self.tx.put::<tables::VerkleTrie>(key, value).unwrap();
+        }
+        self.tx.commit().unwrap();
+
+        let num_items = self.batch.num_items();
+        println!(
+            "write to batch time: {}, item count : {}",
+            now.elapsed().as_millis(),
+            num_items
+        );
+
+        self.batch.clear();
+    }
+}
+
+impl<'a, TX: DbTx> ReadOnlyHigherDb for VerkleDb<'a, TX> {
+    fn get_leaf(&self, key: [u8; 32]) -> Option<[u8; 32]> {
+        // First try to get it from cache
+        if let Some(val) = self.cache.get_leaf(key) {
+            return Some(val);
+        }
+        // Now try to get it from batch
+        if let Some(val) = self.batch.get_leaf(key) {
+            return Some(val);
+        }
+        // Now try the disk
+        let mut trie_cursor = self.tx.cursor_read::<tables::VerkleTrie>().ok()?;
+        let mut labelled_key = Vec::with_capacity(key.len() + 1);
+        labelled_key.push(LEAF_TABLE_MARKER);
+        labelled_key.extend_from_slice(&key);
+        trie_cursor.seek_exact(labelled_key)
+            .ok()
+            .and_then(|result| result)
+            .and_then(|(_, value)| value.try_into().ok())
```

Comment:
This part is redundant in all impls of verkleDb, ig need to add functionality such that, to avoid multiple cursor creation